core/client/src/client.rs

```diff
@@ -1179,6 +1179,48 @@ impl<B, E, Block, RA> Client<B, E, Block, RA> where
 		Ok(None)
 	}
 
+	fn get_descendants(&self, target: Block::Hash, maybe_skip: Option<Block::Hash>) -> error::Result<Vec<Block::Hash>> {
+		let mut children = self.backend.blockchain().children(target)?;
+		if let Some(skip) = maybe_skip {
+			children.retain(|&c| c != skip);
+		}
+		let mut descendants = Vec::new();
+		for child in children {
+			descendants.push(child);
+			let d = self.get_descendants(child, None)?;
+			descendants.extend(d);
+		}
+		Ok(descendants)
+	}
+
+	/// Gets the uncles of the block with `target_hash` going back `max_generation` ancestors.
+	pub fn uncles(&self, target_hash: Block::Hash, max_generation: NumberFor<Block>)
+		-> error::Result<Vec<Block::Hash>> {
+		let load_header = |id: Block::Hash| -> error::Result<Block::Header> {
+			match self.backend.blockchain().header(BlockId::Hash(id))? {
+				Some(hdr) => Ok(hdr),
+				None => Err(ErrorKind::UnknownBlock(format!("Unknown block {:?}", id)).into()),
+			}
+		};
+
+		let genesis_hash = self.backend.blockchain().info().unwrap().genesis_hash;
+		let genesis = load_header(genesis_hash)?;
+		let mut current = load_header(target_hash)?;
+		let mut uncles = Vec::new();
+
+		if genesis == current { return Ok(uncles); }
+		let mut ancestor = load_header(*current.parent_hash())?;
+
+		for _generation in 0..max_generation.as_() {
+			uncles.extend(self.get_descendants(ancestor.hash(), Some(current.hash()))?);
+			current = ancestor;
+			if genesis == current { break; }
```

Comment:
just check if current `hash == genesis hash` and don't load the genesis header. and you can avoid doing the hash calculations in all cases but the first by book-keeping the last `parent_hash`.